src/tribler/core/components/libtorrent/download_manager/download_manager.py

```diff
@@ -258,17 +257,16 @@ def create_session(self, hops=0, store_listen_port=True):
         else:
             ltsession = lt.session(lt.fingerprint(*fingerprint), flags=0) if hops == 0 else lt.session(flags=0)
 
-        libtorrent_port = self.config.port or default_network_utils.get_random_free_port()
-        self._libtorrent_port = libtorrent_port
-        self._logger.info(f'Libtorrent port: {libtorrent_port}')
         if hops == 0:
             settings['user_agent'] = 'Tribler/' + version_id
             enable_utp = self.config.utp
             settings['enable_outgoing_utp'] = enable_utp
             settings['enable_incoming_utp'] = enable_utp
 
             settings['prefer_rc4'] = True
-            settings["listen_interfaces"] = "0.0.0.0:%d" % libtorrent_port
+
+            # We do not specify the port in settings dict as we allow libtorrent to automatically choose the port
+            # settings["listen_interfaces"] = "0.0.0.0:%d" % libtorrent_port
```

Comment:
Allowing users to specify a port for torrent traffic is [critical functionality for some users](https://www.downloadprivacy.com/how-to-torrent/port-forwarding). The other ports probably don't matter as much but I'd leave the libtorrent port config option as-is.

Reply:
Thank you for this link!

Thinking about this, the current Tribler logic regarding the libtorrent port number handling does not look correct to me. This is how, in my understanding, the current logic of `DownloadManager.create_session()` works:

0. Initially, [in LibtorrentSettings](https://github.com/Tribler/tribler/blob/24d189c09a88bd91379034f8f66659ec9a6a0676/src/tribler/core/components/libtorrent/settings.py#L25), the port value is `None` by default (and a user can set a specific value):
```python
class LibtorrentSettings(TriblerConfigSection):
    ...
    port: Optional[int] = None
    ...
```
1. The value is read from config, and [an empty value is replaced with a random free port](https://github.com/Tribler/tribler/blob/090ae756e76fd3649da553cd1ab5bab9671d3f2f/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L261):
```python
libtorrent_port = self.config.port or default_network_utils.get_random_free_port()
```
2. The resulting port value [is passed to libtorrent](https://github.com/Tribler/tribler/blob/090ae756e76fd3649da553cd1ab5bab9671d3f2f/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L289):
```python
settings["listen_interfaces"] = "0.0.0.0:%d" % libtorrent_port
...
self.set_session_settings(ltsession, settings)
```
3. The libtorrent [is requested to listen on a port in a range](https://github.com/Tribler/tribler/blob/090ae756e76fd3649da553cd1ab5bab9671d3f2f/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L303), starting from a specified port value:
```python
ltsession.listen_on(libtorrent_port, libtorrent_port + 10)
```
4. Then, if the actual port used by libtorrent is different from the port that was initially specified, [the actual port value is stored in the config](https://github.com/Tribler/tribler/blob/090ae756e76fd3649da553cd1ab5bab9671d3f2f/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L305):
```python
if libtorrent_port != ltsession.listen_port() and store_listen_port:
    self.config.port = ltsession.listen_port()
```
(the `store_listen_port` parameter value is True by default and [is not specified](https://github.com/Tribler/tribler/blob/090ae756e76fd3649da553cd1ab5bab9671d3f2f/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L340) when the `DownloadManager.create_session` method is called).

From this, my conclusion is the following: even when a user specifies the libtorrent port value in settings, it is possible that the value will be changed to a different value (at step 3) if the original port is already occupied. Then the value provided by a user will be overridden in settings by a new value without any confirmation by a user.

From what I currently understand, the correct logic should work in the following way:
1. If the port value is explicitly specified by a user, it should be used. If the specified port is already occupied, Tribler should not use a different port silently. Instead, it should notify the user that the specified port is unavailable and stop.
2. If the value is not specified, Tribler should allow libtorrent to choose the port automatically. In that case, Tribler should not store the randomly assigned port value in `LibtorrentSettings`, as it is not guaranteed that the same port will be available next time, and it is not important to use the same port next time. Only the port value explicitly specified by a user should go to LibtorrentSettings.

@qstokkink what do you think?

Reply:
Sounds good 👍 That's what I would expect from my torrent client.

Reply:
After the offline discussion with @xoriole we decided to extract libtorrent port handling changes to a separate PR